### PR TITLE
IGNITE-14657 README.md for ignite-configuration module.

### PR DESCRIPTION
**Pull request by ibessonov**
Signed-off-by: ibessonov <bessonov.ip@gmail.com>

### CODE REVIEW

File: modules/configuration/README.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+### Configuration Schema
+
+Type-safe schema of a configuration, which is used for generating public API interfaces and
+internal implementations to avoid writing boilerplate code. 
+
+All schema classes must end with the `ConfigurationSchema` suffix.
+
+### Configuration Registry
+
+`ConfigurationRegistry` is the entry point of the module.
```
Comment (sashapolo) [inline modules/configuration/README.md:19]: Add more text here

File: modules/configuration/README.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+this interface are generated automatically and are mandatory for registering the configuration roots.
+
+### Example Schema
+
+An example configuration schema may look like the following:
+
+```java
+@ConfigurationRoot(rootName = "root", type = ConfigurationType.LOCAL)
+public static class ParentConfigurationSchema {
+    @NamedConfigValue
```
Comment (sashapolo) [inline modules/configuration/README.md:36]: Add an explanation about `NamedConfigValue` and `NamedLists`

File: modules/configuration/README.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    
+    @Value
+    @Immutable
+    public String str;
+}
+```
+
+* `@ConfigurationRoot` marks the root schema. It contains the following properties:
+  * `type` property, which can either be `LOCAL` or `DISTRIBUTED`. This property dictates the _storage_ type used 
+    to persist the schema — `Vault` or `Metastorage`;
```
Comment (sashapolo) [inline modules/configuration/README.md:56]: What is the `Vault` and `Metastorage` and what are the consequences of choosing one?

File: modules/configuration/README.md [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+  * All Ignite configuration instances can be represented by a forest, where every node has a name, usually referred 
+    to as a _key_. The `rootName` property assigns a _key_ to the root node of the tree that will represent 
+    the corresponding configuration schema;
+* `@Config` is similar to the `@ConfigurationRoot` but represents an inner configuration node;
+* `@ConfigValue` marks a nested schema field. Cyclic dependencies are not allowed;
+* `@NamedConfigValue` is similar to `@ConfigValue` but such fields will basically become `Map<String, declared_schema>`;
```
Comment (sashapolo) [inline modules/configuration/README.md:61]: > but such fields will basically become `Map<String, declared_schema>`;

This sentence makes little sense

File: modules/configuration/README.md (hunk 1 above)
Comment (ibessonov) [reply modules/configuration/README.md]: Done

File: modules/configuration/README.md (hunk 2 above)
Comment (ibessonov) [reply modules/configuration/README.md:37]: Done

File: modules/configuration/README.md [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+Type-safe schema of a configuration, which is used for generating public API interfaces and
+internal implementations to avoid writing boilerplate code. 
+
+All schema classes must end with the `ConfigurationSchema` suffix.
+
+### Configuration Registry
+
+`ConfigurationRegistry` is the entry point of the module. It is used to register root keys, validators, storages and to
+start / stop component. Please refer to the class for more details.
```
Comment (sashapolo) [inline modules/configuration/README.md:20]:
```suggestion
start / stop the component. Refer to the class javadocs for more details.
```

File: modules/configuration/README.md [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    
+    @Value
+    @Immutable
+    public String str2;
+}
+```
+
+* `@ConfigurationRoot` marks the root schema. It contains the following properties:
+  * `type` property, which can either be `LOCAL` or `DISTRIBUTED`. This property dictates the _storage_ type used 
+    to persist the schema — `Vault` or `Metastorage`. `Vault` stores data locally while `Metastorage` is a distributed
```
Comment (sashapolo) [inline modules/configuration/README.md:57]:
```suggestion
    to persist the schema — `Vault` or `Metastorage`. `Vault` stores data locally, while `Metastorage` is a distributed
```

File: modules/configuration/README.md [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+    system that should store only cluster-wide configuration properties;
+  * `rootName` property assigns a _key_ to the root node of the tree that will represent 
+    the corresponding configuration schema;
+* `@Config` is similar to the `@ConfigurationRoot` but represents an inner configuration node;
+* `@ConfigValue` marks a nested schema field. Cyclic dependencies are not allowed;
+* `@NamedConfigValue` is similar to `@ConfigValue` but such fields represent collection of properties, not a single
```
Comment (sashapolo) [inline modules/configuration/README.md:63]:
```suggestion
* `@NamedConfigValue` is similar to `@ConfigValue`, but such fields represent a collection of properties, not a single
```

File: modules/configuration/README.md [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+  instance. Every element of the collection will have a `String` name, which makes it pretty much like a `Map`.
```
Comment (sashapolo) [inline modules/configuration/README.md:64]:
```suggestion
  instance. Every element of the collection will have a `String` name, similar to a `Map`.
```

File: modules/configuration/README.md [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+  `NamedListConfiguration` interface is used to represent this field in generated configuration classes.
```
Comment (sashapolo) [inline modules/configuration/README.md:65]:
```suggestion
  `NamedListConfiguration` interface is used to represent this field in the generated configuration classes. 
```

File: modules/configuration/README.md [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+* `@Value` annotation marks the _leaf_ values. `hasDefault` property can be used to set default values for fields:
+  if set to `true`, the default value will be used to initialize the annotated configuration field in case no value 
+  has been provided explicitly. This annotation can only be present on fields of the following types:
+  * `boolean` or `boolean[]`
+  * `int` or `int[]`
+  * `long` or `long[]`
+  * `double` or `double[]`
+  * `String` or `String[]`
+    
+  All _leaves_ must be public and corresponding configuration values **can't be null**;
```
Comment (sashapolo) [inline modules/configuration/README.md:75]:
```suggestion
  All _leaves_ must be public and corresponding configuration values **must not be null**;
```

**Review by sashapolo — APPROVED** [review]

**Pull request merged**